server/proxy/notify.go

```diff
@@ -15,167 +17,199 @@ import (
-// "user_left_channel" and "bot_left_channel" notifications to subscribed apps.
-func (p *Proxy) NotifyUserLeftChannel(channelID, userID string) {
-	p.notifyUserChannel(channelID, userID, false, "NotifyUserLeftChannel")
-}
+	log := p.conf.NewBaseLogger().With("subject", subject)
 
-func (p *Proxy) notifyUserChannel(channelID, userID string, joined bool, method string) {
 	mm := p.conf.MattermostAPI()
-	log := p.conf.NewBaseLogger().With("method", method)
-	user, err := mm.User.Get(userID)
+	user, err := mm.User.Get(member.UserId)
 	if err != nil {
 		log.WithError(err).Debugf("failed to get user")
 		return
 	}
-	channel, err := mm.Channel.Get(channelID)
+	channel, err := mm.Channel.Get(member.ChannelId)
 	if err != nil {
-		log.WithError(err).Debugf("%s: failed to get channel", method)
+		log.WithError(err).Debugf("%s: failed to get channel", subject)
 		return
 	}
 
-	subject := apps.SubjectUserJoinedChannel
-	if !joined {
-		subject = apps.SubjectUserLeftChannel
-	}
-	p.notify(
-		nil,
+	// Notify on user_joined|left_channel subscriptions specific to the channel
+	// that may include any user.
+	p.notifyAll(
 		apps.Event{
 			Subject:   subject,
-			ChannelID: channelID,
+			ChannelID: channel.Id,
 		},
 		apps.UserAgentContext{
-			ChannelID: channelID,
+			ChannelID: channel.Id,
 			TeamID:    channel.TeamId,
 			UserID:    user.Id,
 		},
+		nil, // no special filtering, notify all subscriptions matching the event.
+		nil, // no special expand rules for "any user" subscriptions.
 	)
-	if !user.IsBot {
-		return
-	}
 
-	// If the user is a bot, process SubjectBot...Channel; only notify the
-	// app with the matching BotUserID.
-	allApps := p.store.App.AsMap(store.EnabledAppsOnly)
-	subject = apps.SubjectBotJoinedChannel
-	if !joined {
-		subject = apps.SubjectBotLeftChannel
-	}
-	p.notify(
-		func(sub store.Subscription) bool {
-			if app, ok := allApps[sub.AppID]; ok {
-				return app.BotUserID == userID
-			}
-			return false
-		},
+	// Notify on "self" subscriptions for the user.
+	p.notifyAll(
 		apps.Event{
 			Subject: subject,
-			TeamID:  channel.TeamId,
 		},
 		apps.UserAgentContext{
-			ChannelID: channelID,
+			ChannelID: channel.Id,
 			TeamID:    channel.TeamId,
-			UserID:    user.Id,
+			UserID:    member.UserId,
+		},
+		// Include subscriptions where the user (added/removed channel member)
+		// is the same as the subscriber.
+		func(sub store.Subscription) bool {
+			return sub.OwnerUserID == member.UserId
 		},
+		// special expand for "self" subscriptions.
+		newExpandSelfGetter(mm, user, member, nil, channel),
 	)
-}
 
-// NotifyUserJoinedTeam handles plugin's UserHasJoinedTeam callback. It emits
-// "user_joined_team" and "bot_joined_team" notifications to subscribed apps.
-func (p *Proxy) NotifyUserJoinedTeam(teamID, userID string) {
-	p.notifyUserTeam(teamID, userID, true, "NotifyUserJoinedTeam")
+	// Notify on the deprecated bot_joined|left_channel subscriptions.
+	if user.IsBot {
```

Comment:
We can save an indentation here with `if !user.IsBot`

Reply:
I'm not sure if that would make the code easier to read. We might add additional code in the future that handles more cases.